MH-12607, Process-Smil - edit and encode to multiple delivery formats

phoebem opened the pull request:
This is the process-smil part of 12607, updated to new composer code.

Commit:
Process-Smil - edit and encode to multiple delivery formats

Commit:
Process-Smil - edit and encode to multiple delivery formats

Commit:
trailing spaces

Commit:
trailing spaces

Commit:
Shorter lines

Commit:
Shorter lines

Commit:
Merge branch 'f/MH-12607-part-2-process-smil' of https://github.com/phoebem/opencast into f/MH-12607-part-2-process-smil

Commit:
Merge branch 'f/MH-12607-part-2-process-smil' of https://github.com/phoebem/opencast into f/MH-12607-part-2-process-smil

Commit:
Merge branch 'f/MH-12607-part-2-process-smil' of https://github.com/phoebem/opencast into f/MH-12607-part-2-process-smil

Commit:
Fixed bad link

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/process-smil-woh.md:
```diff
@@ -0,0 +1,131 @@
+of the encoding profile command.
+1.  Care must be taken that **no complex filters** are used in the encoding profiles used for this workflow,
+as it can cause a conflict and ffmpeg will fail.
+2.  Encoded target recording are distinguished by the suffix, it is important that **all the encoding profiles
```
The sentence seems weird

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/process-smil-woh.md:
```diff
@@ -0,0 +1,131 @@
+3.  If using this with the editor, be sure to set the "create-smil-skip-processing" key in the editor to true.
+
+
+## Example:
```
I think the text here should be moved down to the Operation Example

Commit:
Fixed documentation

phoebem commented:
I rewrote parts of it and moved the notes down, hopefully it is clearer.

Thanks, Phoebe.

> On May 26, 2018, at 10:02 AM, Sven Stauber <notifications@github.com> wrote:
> 
> @staubesv commented on this pull request.
> 
> In docs/guides/admin/docs/workflowoperationhandlers/process-smil-woh.md <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23discussion-5Fr191049570&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=RGJut2dfzrEMOj8p-xgNvk03Ro4ccGzIfZhfRLhWpbM&s=XwRw3eVDrYnD0BDv0_XvAgzkpBsNKR0LgAHbag9OOh4&e=>:
> 
> > +are taken from the corresponding sections.
> +
> +Operations on each source section can run on different servers concurrently.
> +Each source section will run one ffmpeg process.
> +
> +### Note:(Very Important)
> +Each source flavor generates all the target formats in one ffmpeg call by incorporating relevant parts
> +of the encoding profile command.
> +1.  Care must be taken that **no complex filters** are used in the encoding profiles used for this workflow,
> +as it can cause a conflict and ffmpeg will fail.
> +2.  Encoded target recording are distinguished by the suffix, it is important that **all the encoding profiles
> +used have distinct suffixes** or the target video tagging can be wrong.
> +3.  If using this with the editor, be sure to set the "create-smil-skip-processing" key in the editor to true.
> +
> +
> +## Example:
> I think the text here should be moved down to the Operation Example
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23pullrequestreview-2D123557733&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=RGJut2dfzrEMOj8p-xgNvk03Ro4ccGzIfZhfRLhWpbM&s=zgqeq3ON0seHG4OEfjN0NR5z8HTOKelLfvf0Y3zyPSA&e=>, or mute the thread <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGPyNKrrk5iFoHnRBLcF4RsfDuGUxBltks5t2WCLgaJpZM4UOpLc&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=RGJut2dfzrEMOj8p-xgNvk03Ro4ccGzIfZhfRLhWpbM&s=YSzpl8OYY5ppiIUpjYZwqz8eSnLMraMXo7mpOvpaL34&e=>.
>

rute-santos committed:
MH-12607-job-load-process-smil Calculate the job load based on encoding profiles used.

rute-santos committed:
MH-12607-job-load-process-smil Calculate the job load based on encoding profiles used.

Commit:
trailing space

phoebem committed:
Merge pull request #1 from harvard-dce/f/MH-12607-job-load-process-smil

MH-12607 Patch: calculate the process-smil job load based on the encoding profiles used.

Commit:
Merge branch 'f/MH-12607-part-2-process-smil' of https://github.com/phoebem/opencast into f/MH-12607-part-2-process-smil

Commit:
line length

Commit:
Merge branch 'f/MH-12607-part-2-process-smil' of https://github.com/phoebem/opencast into f/MH-12607-part-2-process-smil

Commit:
spaces

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/editor-woh.md:
```diff
@@ -17,6 +17,7 @@ distributed systems) to be installed.
 |target-flavor-subtype|`trimmed`|The flavor subtype to be applied to all resulting videos, e.g. for a value of `baz`, a track with flavor `foo/bar` will generate another track with flavor `foo/baz`|
 |target-smil-flavor| `smil/cutting` |the flavor of the SMIL file containing the final video segments.<br/>Should be the same as the `smil.catalog.flavor` property in `etc/org.opencastproject.adminui.cfg`|
 |skip-if-not-trimmed|`false`       |(Optional) if set to `true`, the track encoding will be skipped if no trimming points were defined (i.e. there is only one segment from the very beginning to the very end of the video). Defaults to `false`|
+|create-smil-skip-processing|`true`|Do not do the actual encoding, just create the smil file and exit. This option is used with *process-smil* workflow operation, which will use the smil to run the encodings then. Default is false. |
```
Suggestion: Just "skip-processing" would be "long enough"

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/process-smil-woh.md:
```diff
@@ -0,0 +1,152 @@
+>presentation/preview is encoded with "mp4-vga-medium" and tagged with "engage-download","rss" and "atom".
+
+
+## Example:
```
Remove trailing colon in title

staubesv replied:
Example
Parameter Table
Operation Example

could be collapsed to

Parameter Table
Operation Example (incl. Example(

staubesv commented on docs/guides/admin/docs/workflowoperationhandlers/process-smil-woh.md:
```diff
@@ -0,0 +1,152 @@
+
+If presenter/work is to be encoded with "mp4-low.http,mp4-medium.http" and
+presentation/work is to be encoded with "mp4-vga-medium,mp4-medium.http",
+and the target media are flavored as "presenter/delivery" and "presentation/delivery" respectively.
```
not final stop here

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -148,9 +164,29 @@
   /** The formatter for load values */
   private static final DecimalFormat df = new DecimalFormat("#.#");
 
+  /** Configuration for process-smil transition duration - TODO: use smil transition element */
```
Maybe just create a JIRA ticket for the "SMIL transition element" feature and remove the TODO from the comment. At the end of the day, source code is not necessarily the optimal place to put feature requests ;-)

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1229,9 +1295,9 @@ protected String process(Job job) throws ServiceRegistryException {
       Track secondTrack;
-      String encodingProfile = arguments.get(0);
-
+      String encodingProfile = arguments.get(0); // rewrite for ProcessSmil
```
Suggestion: Remove comment as it seems not helpful

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,283 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+  /**
+   * Fetch specified or first SmilMediaParamGroup from smil
+   *
+   * @param smil
```
descriptions missing

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,283 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+
+    SmilMediaParamGroup trackParamGroup;
+    ArrayList<String> inputfile = new ArrayList<String>();
+    Map<String, String> props = new HashMap<String, String>();
```
Some people might be tempted to change that into new HashMap<>(); later on ;-)

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,283 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+      try {
+        outputs = encoderEngine.multiTrimConcat(inputs, edits, profiles, transitionDuration, !"a".equals(mediaType),
+                !"v".equals(mediaType));
+        // TODO: transitionDuration per transition from editor
```
Maybe remove comment and create JIRA issue for this

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +462,530 @@ private void handleEncoderOutput(List<File> output, String message) {
+   * @return
+   */
+  private static List<VideoClip> sortSegments(List<VideoClip> edits, double gap) {
+    // TODO: add transition to each edited clip in place of transitionDuration
```
JIRA instead of comment

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +462,530 @@ private void handleEncoderOutput(List<File> output, String message) {
+    return clips;
+  }
+
+  // TODO: add transition to each edited clip in place of transitionDuration
```
JIRA instead of comment

staubesv commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +462,530 @@ private void handleEncoderOutput(List<File> output, String message) {
+   * @param clips
+   *          - video segments as indices into the media files by time
+   * @param transitionDuration
+   *          - length of transition in MS between each segment, TODO: transitionDuration to be replaced by transIn and
```
JIRA instead of comment

staubesv commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/ComposerService.java:
```diff
@@ -258,4 +260,28 @@ Job image(Track sourceTrack, String profileId, Map<String, String> properties) t
+   * Reads a smil definition and create one media track in multiple delivery formats. The track in the smil is selected
+   * by "trackParamGroupId" which is the paramGroup in the smil The multiple delivery formats are determined by a list
+   * of encoding profiles by name. The resultant tracks will be tagged by profile name. The smil file can contain more
+   * than one source track but they must have the same dimension. This is used mainly on smil.xml from the editor. There
```
Remove "There is a 2s fadein/fadeout between each clip" or replace with correct comment (the transitionDuration is configurable)

staubesv commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/ComposerService.java:
```diff
@@ -258,4 +260,28 @@ Job image(Track sourceTrack, String profileId, Map<String, String> properties) t
+   *
+   * @param smil
+   *          - Describes one media (can contain multiple source in ws) and editing instructions (in out points) for
+   *          conte
```
conte?

staubesv commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/VideoClip.java:
```diff
@@ -0,0 +1,106 @@
+
+import java.util.Comparator;
+
+/* #DCE OPC-29
```
?

staubesv commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/VideoClip.java:
```diff
@@ -0,0 +1,106 @@
+ * srcId is a bit redundant, but needed for convenience.
+ * By default, it is all video
+ */
+// #DCE added natural ordering by start time
```
#DCE...?

staubesv commented on modules/composer-service-api/src/main/java/org/opencastproject/composer/api/VideoClip.java:
```diff
@@ -0,0 +1,106 @@
+    return "VideoClip [srcId=" + srcId + ", start=" + start + ", end=" + end + "]";
+  }
+
+  // [MATT-2095-stat-s-139-lecture-2-error-res] #DCE - Natural order of the clips is by start time
```
Weird comment...

staubesv commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ProcessSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,728 @@
+      this.paramGroupId = id;
+    }
+
+    @SuppressWarnings("unused")
```
Is it really not used? Why include it then?

staubesv commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ProcessSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,728 @@
+          hasAudio = false;
+      }
+      if (!hasVideo)
+        mediaType = "a";
```
Would be preferable to have constants for "a", "v", and "av"...

staubesv commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ProcessSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,728 @@
+   * parse all the encoding jobs to collect all the composed tracks, if any of them fail, just fail the whole thing and
+   * try to clean up
+   *
+   * @param encodingJobs
```
Missing descriptions

staubesv commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ProcessSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,728 @@
+  /**
+   * return true if trackFlavor matches sourceFlavor
+   *
+   * @param trackFlavor
```
missing description

staubesv commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ProcessSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,728 @@
+   * Given a mediaPackage, a smil and a srcFlavor with wild card, this will return a structure of smil groups, each with
+   * a single flavor and mp tracks for that flavor only
+   *
+   * @param mediaPackage
```
Missing description

staubesv commented:
Hi @phoebem Note that I will require multiple passes for this review as the PR is quite large.

I've just finished a first "quick skim-over" and identified a number of "issues":

- For a number of methods, the description of the parameters is missing. I've commented this for some methods but not for all methods. Please either don't specify parameters (@param) not at all or provide a description. Just documenting the parameter names seems useless.

- It doesn't seem particularly useful to have "TODO" comments in the source code - such information should usually be tracked in the issue tracker instead. 

- You often use code like "Array\<String> = new Array\<String>;" While this is correct, I've seen a number of people that tend to change this into "Array\<String> = new Array\<>;". It seems to be the preferred notation and we could avoid those "little changes" by just using that notation right at the beginning

- The values "a", "v" and "av" seem worth the introduction of constants as used in several code locations

- Some other suggestions for improvements (see comments)

Note that those are just suggestions - feel free to object in case of disagreement.

I suggest that I wait with the next review round until the "issues" above are fixed.

Is that ok for you?

Commit:
Added constants for audio/video only, fixed comments

Commit:
shortened param name + punctuation fix

Commit:
Updated java syntax

phoebem commented:
Hi Sven,

Thanks for the comments on this huge PR.

> On Jun 1, 2018, at 5:47 AM, Sven Stauber <notifications@github.com> wrote:
> 
> Hi @phoebem <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_phoebem&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=4OBxZn6rQ1ict6p6Ud1LcsL9zS2DTdNM1az4GnE9tXU&s=m6ZiWfSDA4JjXod9k41s1TRtjrQs8vMYoC-bXtX7Bkg&e=> Note that I will require multiple passes for this review as the PR is quite large.
> 
> I've just finished a first "quick skim-over" and identified a number of "issues":
> 
> For a number of methods, the description of the parameters is missing. I've commented this for some methods but not for all methods. Please either don't specify parameters (@param <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_param&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=4OBxZn6rQ1ict6p6Ud1LcsL9zS2DTdNM1az4GnE9tXU&s=1bDigKlvsTCuIaNSK3xteZvLwoJ4vCP5plWWhT6xMB4&e=>) not at all or provide a description. Just having the parameter name seems useless.
> 
> It doesn't seem particularly useful to have "TODO" comments in the source code - such information should usually be tracked in the issue tracker instead.
> 
> You often use code like "Array = new Array;" While this is correct, I've seen a number of people that tend to change this into "Array = new Array<>;". It seems to be the preferred notation and we could avoid those "little changes" by just using that notation right at the beginning
> 
Most of this code was written on jdk 6 before the new syntax. I have just updated as many as I could find.
> The values "a", "v" and "av" seem worth the introduction of constants as used in several code locaStions
> 
I used “a” and “v” originally because I didn’t want to change ComposeService api. I have added them now
> 
> Some other suggestions for improvements (see comments)
> 
> Note that those are just suggestions - feel free to object in case of disagreement.
> 
> I suggest that I wait with the next review round until the "issues" above are fixed.
> 
> Is that ok for you?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23issuecomment-2D393831010&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=4OBxZn6rQ1ict6p6Ud1LcsL9zS2DTdNM1az4GnE9tXU&s=q2TM4Cyx7H_nNcqU1v5s6VQtNarEbsAzw3BZjWV7p48&e=>, or mute the thread <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGPyNAtGfaa27dIitoiDUc1-5FsJiq5xWXks5t4Q2kgaJpZM4UOpLc&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=4OBxZn6rQ1ict6p6Ud1LcsL9zS2DTdNM1az4GnE9tXU&s=4_97aYN44DQejPpwbOl7VkEMiiP2cqYi2-E1tzeSueI&e=>.
>

staubesv commented on etc/workflows/partial-publish-process-smil.xml:
```diff
@@ -0,0 +1,300 @@
+        <configuration key="target-flavor-subtype">trimmed</configuration>
+        <configuration key="interactive">false</configuration>
+        <!-- skip the built in editor transcode to use process-smil-->
+        <configuration key="create-smil-skip-processing">true</configuration>
```
@phoebem skip-processing (instead of create-smil-skip-processing)

staubesv commented on etc/workflows/partial-publish-process-smil.xml:
```diff
@@ -0,0 +1,300 @@
+      </configurations>
+    </operation>
+
+    <!-- Archive smil/cutting right now information to ensure we don't loose it in case later operations fail -->
```
Hmm... is that necessary at all (in means of does WOH editor generate a smil/cutting that needs to be archived)? If yes, you might want to move those operations (tag & snapshot) directly after WOH editor.

staubesv reviewed (changes requested):
Hi @phoebem While testing this, I've realised the WOH editor still uses the configuration key "create-smil-and-skip-processing". Also, WOH snapshot seems not placed optimally.

Could you please fix those two issues?

Commit:
Updated param

staubesv commented on etc/org.opencastproject.composer.impl.ComposerServiceImpl.cfg:
```diff
@@ -0,0 +1,18 @@
+# If it is set to 0, there is no transition between each segment
+# By default, it is a 2 second fade-to-black and fade-from-black and audio fading between each segment
+
+org.composer.process_smil.edit.transition.duration = 2
```
Please adapt to the same style as the configuration properties above, i.e. don't explicitly set the default value, same style of documentation of default value ("# Default: 2)

staubesv replied:
Hi @phoebem 

I've found the following problem while testing:

Step 1: A video is cutted...

![screen shot 2018-06-13 at 14 53 45](https://user-images.githubusercontent.com/1590263/41353574-e92dd3e0-6f1c-11e8-81ad-b40e7bef0c3b.png)

Step 2: To verify that the standard workflow does not break things, it is started (Video Editor->Publish). After the workflow has processed the video, things look like:

![screen shot 2018-06-13 at 15 09 51](https://user-images.githubusercontent.com/1590263/41353611-04dcb002-6f1d-11e8-96fe-8ecb4c2116e0.png)

Problem: If the new process-smil workflow is started from the video editor, processing more or less works (the result is cutted, but the Engage player still shows the full length of the video), but also, when opening the video editor again, things look like:

![screen shot 2018-06-13 at 15 16 12](https://user-images.githubusercontent.com/1590263/41353667-2e777b0e-6f1d-11e8-9086-8092a31c7ae4.png)


I have verified that the segments in the timeline don't get messed up in opencast/develop or in this branch when not using the new process-smil workflow.

Can you please have a look at this issue?

staubesv reviewed (changes requested):
Hi phoebem

Besides the new minor comment (default values), it seems overkill to add separate workflows (and the amount of duplication this introduces) for the case no theme has been set.
In particular, it seems weird to ask an user what workflow to start in the video editor: If the user has selected a theme, he might actually want this to be rendered, right?

Note that this can be addressed by calling WOH theme which will set the workflow instance variable "theme_active" to "true" if the event has a theme and "false" otherwise. Based on this information, Opencast can determine (without user interaction) what workflow operations are needed.

Can you please restructure the workflows so that the process-smil optimizaiton is used if theme_active is "false" but theme things (bumper, trailer, watermark, title slide) are rendered in case a theme is active? Please try to avoid too much duplication in the workflow definitions.

That way, the user does not need to decide what workflow to start but Opencast can still apply optimisations where appropriate. And we can avoid quite some duplication in workflow definitions.

phoebem commented:
This sample workflow is provided as an simple example on how to use this handler, 
so that users can write their own workflows. I purposefully avoided changing the default workflow.

The encoding profiles are also declared in the process-smil workflow operation and
not selected from the menu choice, so there is another deviation from the default workflow.

Are you suggesting that instead of a separate sample workflow as a choice to users, 
you want to have to one option after trim, so that partial_publish.xml will exercise either old and new code based on 
  if=“(NOT ${theme_active})"

Given that the workflow has to be edited to use themes, perhaps we can make partial_publish.xml call “theme” to set the variable first, then
it can include either partial_publish_process-smil or partial_publish_themed, and then calls the publish-* workflows.
What do you think?

Phoebe.

> On Jun 7, 2018, at 7:04 AM, Sven Stauber <notifications@github.com> wrote:
> 
> @staubesv requested changes on this pull request.
> 
> Hi phoebem
> 
> Besides the new minor comment (default values), it seems overkill to add separate workflows (and the amount of duplication this introduces) for the case no theme has been set.
> In particular, it seems weird to ask an user what workflow to start in the video editor: If the user has selected a theme, he might actually want this to be rendered, right?
> 
> Note that this can be addressed by calling WOH theme which will set the workflow instance variable "theme_active" to "true" if the event has a theme and "false" otherwise. Based on this information, Opencast can determine (without user interaction) what workflow operations are needed.
> 
> Can you please restructure the workflows so that the process-smil optimizaiton is used if theme_active is "false" but theme things (bumper, trailer, watermark, title slide) are rendered in case a theme is active?
> 
> That way, the user does not need to decide what workflow to start and we can avoid quite some duplication in workflow definition.
> 
> In etc/org.opencastproject.composer.impl.ComposerServiceImpl.cfg <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23discussion-5Fr193705528&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Oh0jcVeqNIgyTygWrlJjRSbJk2r_uuaL4xcurh2i3_Q&s=ULBgdlmNRCXr23p8k2_ANGkdx9LQCQ2HpMcbtm6plwU&e=>:
> 
> > +# The factor represents the adjustment that should be made assuming each profile job load was specified
> +# based on it running with one input -> one output so normally will be a number 0 < n <= 1
> +# Their job loads maxes out at job.load.max.multiple.profiles.
> +# Default: 0.8
> +#job.load.max.multiple.profiles=0.8
> +
> +# The process-smiltrack operation stitches all the editor segments and transcode with multiple
> +# encoding profiles concurrently.
> +# Default: 0.5
> +#job.load.factor.process.smil=0.5
> +
> +# Transition duration in seconds between each edited section when using process-smiltrack operation
> +# If it is set to 0, there is no transition between each segment
> +# By default, it is a 2 second fade-to-black and fade-from-black and audio fading between each segment
> +
> +org.composer.process_smil.edit.transition.duration = 2
> Please adapt to the same style as the configuration properties above, i.e. don't explicitly set the default value, same style of documentation of default value ("# Default: 2)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23pullrequestreview-2D126729273&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Oh0jcVeqNIgyTygWrlJjRSbJk2r_uuaL4xcurh2i3_Q&s=X58WCmbDECAYzHgx8SdnowEWduQAzcMave8_bPfQCao&e=>, or mute the thread <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGPyNJYMCJu8svW2LqCUtE893W6Ka-2DB3ks5t6QizgaJpZM4UOpLc&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Oh0jcVeqNIgyTygWrlJjRSbJk2r_uuaL4xcurh2i3_Q&s=FN_5n66aN4Bt-MfghaSBA4A7dbEg86vEUUQyLx-JYPY&e=>.
>

staubesv commented:
Oh sorry, I haven't realised that it also does three different formats that can't be configured using the configuration panel. In that case, applying the optimisation based on the variable theme_active obviously won't work.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,284 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+        throw new EncoderException("Unable to create a job - Exception in Parsing Smil", e);
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
```
Don't print stacktraces

phoebem replied:
Uggh! I thought I removed them all. Thanks for catching that.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,284 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+        }
+      } catch (ServiceRegistryException e) {
+        throw new EncoderException("Unable to create a job", e);
+      } catch (Exception e) {
```
Why catch the Exception if the next line catches it anyway?

phoebem replied:
So that it is obvious from the log why it failed, in this case, whether there is an error in parsing the smil file.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,284 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+        + " but media is Video Only");
+      } else if (AUDIO_ONLY.equals(mediaType) && profile.getApplicableMediaType() == EncodingProfile.MediaType.Visual) {
+        logger.warn("Profile '" + profileId1 + "' supports " + profile.getApplicableMediaType()
+        + " but media is Audio Only");
```
```Java
logger.warn("Profile '{}' supports {} but media is Audio Only", profileId1, profile.getApplicableMediaType());
```

phoebem replied:
The new syntax is so pythonic! I updated it.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java:
```diff
@@ -1857,4 +1935,284 @@ public Job demux(Track sourceTrack, String profileId) throws EncoderException, M
+                      inputfile.add(clipSourceFile.getAbsolutePath()); // add track
+                      props.put("in.video.path" + index, sourceFile.getAbsolutePath());
+                      index = inputfile.indexOf(clipSourceFile.getAbsolutePath());
+                      logger.info("Adding file {} " + index, inputfile);
```
Weird log

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+      }
+      return process(command); // Run the ffmpeg command
+    } catch (Exception e) {
+      //logger.error("MultiTrimConcat failed to run command {} ", e.getMessage());
```
…

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+          logger.info(pad + matcher.toString());
+          return matcher.group(1);
+        }
+      } catch (Exception e) {
```
The code in the `try` block should not throw any exceptions, should it?

phoebem replied:
It does when pad is null.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+      final Pattern outpad = Pattern.compile("\\[(\\d+:[av\\d{1,2}])\\]");
+      try {
+        Matcher matcher = outpad.matcher(pad); // throws exception if pad is null
+        if (matcher != null && matcher.matches()) {
```
`matcher` is never `null`

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+            i++;
+          } else if (opt.startsWith("-filter_complex") || opt.startsWith("-lavfi")) { // safer to quit now than to
+            // baffle users with strange errors later
+            i++;
```
Never used since an exception is thrown

phoebem replied:
The exception is thrown only if ffmpgCmd cannot be parsed and tokenized.

lkiesow commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -392,4 +461,526 @@ private void handleEncoderOutput(List<File> output, String message) {
+        clauses.add(outmaps.getVideoFilter());
+      }
+      while (clauses.remove(null)) {
+      } // remove all empty filters
```
Not very important but maybe this is a bit cleaner:
```Java
clauses.removeIf(Objects::isNull);
```

phoebem replied:
Updated

lkiesow reviewed (changes requested):
Swiftly read through the code. There are a few minor things I found but nothing major.

staubesv commented:
@phoebem I suggest you have a look at Lars' comments. After that, I will run final tests and continue with merging this PR.

Commit:
Updates for PR comments

Commit:
Removed archive tag

phoebem commented:
Hi Sven,

I found the problem with the editor. It is the archive tag. 
The editor was confused by additional media asset and could not determine the length of the original video.
After I removed it, re-editing shows the original video + trim points from the last edit.

As for the engage player playing the full length video, I couldn’t reproduce it. 
I accessed the player from both the circle-arrow button as well as from "event details > publication" tab
 and engage played the edited video every time. 
In all my test cases, the engage video was trimmed correctly after the edit.
Can you send me a detailed set of steps to reproduce the problem?

Here are my test cases, let me know if I missed anything, the source videos are either 1 min or 15 min.
single presenter video - hold -  edit - process-smil, re-edit - proess-smil
single presenter video - publish - edit  - process-smil
single presentation video - hold - edit - process-smil
single presentation video, no audio - publish - edit - process-smil  - re-edit - process-smil
presenter/presentation - hold - edit - process-smil, re-edit - process-smil

I ran 2 standard workflows as reference.
single presenter video - hold -  edit - publish
single presentation video - hold - edit - publish

Thanks,
Phoebe.

> On Jun 13, 2018, at 9:20 AM, Sven Stauber <notifications@github.com> wrote:
> 
> @staubesv requested changes on this pull request.
> 
> In etc/org.opencastproject.composer.impl.ComposerServiceImpl.cfg <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23discussion-5Fr195077724&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=JJTBCERHZrxus8J9RoraVVdO2EtK7sB9edXc0JeAmV8&e=>:
> 
> > +# The factor represents the adjustment that should be made assuming each profile job load was specified
> +# based on it running with one input -> one output so normally will be a number 0 < n <= 1
> +# Their job loads maxes out at job.load.max.multiple.profiles.
> +# Default: 0.8
> +#job.load.max.multiple.profiles=0.8
> +
> +# The process-smiltrack operation stitches all the editor segments and transcode with multiple
> +# encoding profiles concurrently.
> +# Default: 0.5
> +#job.load.factor.process.smil=0.5
> +
> +# Transition duration in seconds between each edited section when using process-smiltrack operation
> +# If it is set to 0, there is no transition between each segment
> +# By default, it is a 2 second fade-to-black and fade-from-black and audio fading between each segment
> +
> +org.composer.process_smil.edit.transition.duration = 2
> Hi @phoebem <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_phoebem&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=740INZ4KkTs6lRgiZACRWRvMafQCFkNmzdmYzacBH0s&e=>
> I've found the following problem while testing:
> 
> Step 1: A video is cutted...
> 
>  <https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_1590263_41353574-2De92dd3e0-2D6f1c-2D11e8-2D81ad-2Db40e7bef0c3b.png&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=uaFcKLJdvVn0otqkpYvOWaFwaXPgm8VhSXTH0upiPhc&e=>
> Step 2: To verify that the standard workflow does not break things, it is started (Video Editor->Publish). After the workflow has processed the video, things look like:
> 
>  <https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_1590263_41353611-2D04dcb002-2D6f1d-2D11e8-2D96fe-2D8ecb4c2116e0.png&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=SLFAn8TAC-Ka6Dp5v-G3JeSYMoNJuEVrJYjxF7hFXtU&e=>
> Problem: If the new process-smil workflow is started from the video editor, processing more or less works (the result is cutted, but the Engage player still shows the full length of the video), but also, when opening the video editor again, things look like:
> 
>  <https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_1590263_41353667-2D2e777b0e-2D6f1d-2D11e8-2D9086-2D8092a31c7ae4.png&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=JA_vRqnHLOUPgdE4vR3mdN22qfhuSnsPrJhxjAifYfE&e=>
> I have verified that the segments in the timeline don't get messed up in opencast/develop or in this branch when not using the new process-smil workflow.
> 
> Can you please have a look at this issue?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_264-23pullrequestreview-2D128371347&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=QSlpTM677TV-EowMCP_gHHgvqOEGwtmLJ8BATKtrih8&e=>, or mute the thread <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGPyNMXeHH9jgf3JMPITC6-5FA2R-2DEEDfXks5t8RGhgaJpZM4UOpLc&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=nH_AfwZsqYmTdXtU_qgUcVPgE6q1IOQtBrxGsHwWZ_8&s=mptBMwQqQdz3GNVxL6PTlcpwpAyrIfGts-p4HYsXTCY&e=>.
>

staubesv reviewed (approved):
Ok, it works now.

staubesv commented:
Thanks @phoebem! 

@lkiesow I went through your comments and concluded the @phoebem has either resolved them or provided a reasonable explanation.
